rust/splinterdb-rs/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+                it,
+                &mut key_out,
+                &mut val_out,
+            );
+            // parse key and value into rust slices
+            (
+                ::std::slice::from_raw_parts(
+                    ::std::mem::transmute(key_out.data),
+                    key_out.length as usize,
+                ),
+                ::std::slice::from_raw_parts(
+                    ::std::mem::transmute(val_out.data),
+                    val_out.length as usize,
+                ),
+            )
+        };
+        let r = IteratorResult { key, value };
+        Ok(Some(r))
+    }
+
+    pub fn get_curr(&self) -> Option<&IteratorResult>
+    {
+        match self.state {
+            None => None,
+            Some(ref r) => Some(r),
+        }
+    }
+
+    #[allow(clippy::should_implement_trait)]
+    pub fn next(&mut self) -> Result<bool>
+    {
+        let can_next = unsafe { splinterdb_sys::splinterdb_iterator_can_next(self._inner) } as i32;
+        if can_next == 0 {
+            let rc = unsafe { splinterdb_sys::splinterdb_iterator_status(self._inner) };
+            as_result(rc)?;
+            return Ok(false);
+        }
+        unsafe { splinterdb_sys::splinterdb_iterator_next(self._inner); }
+
+        self.state = Self::_get_current(self._inner)?;
+        Ok(true)
+    }
+
+    #[allow(clippy::should_implement_trait)]
+    pub fn prev(&mut self) -> Result<bool>
+    {
+        let can_prev = unsafe { splinterdb_sys::splinterdb_iterator_can_prev(self._inner) } as i32;
+        if can_prev == 0 {
+            let rc = unsafe { splinterdb_sys::splinterdb_iterator_status(self._inner) };
+            as_result(rc)?;
+            return Ok(false);
+        }
+        unsafe { splinterdb_sys::splinterdb_iterator_prev(self._inner); }
+
+        self.state = Self::_get_current(self._inner)?;
+        Ok(true)
+    }
+}
+
+fn path_as_cstring<P: AsRef<Path>>(path: P) -> std::ffi::CString
+{
+    let as_os_str = path.as_ref().as_os_str();
+    let as_str = as_os_str.to_str().unwrap();
+    std::ffi::CString::new(as_str).unwrap()
+}
+
+impl SplinterDB
+{
+    // Create a new SplinterDB object. This is uninitialized.
+    pub fn new<T: rust_cfg::SdbRustDataFuncs>() -> SplinterDB
+    {
+        SplinterDB {
+            _inner: std::ptr::null_mut(),
+            sdb_cfg: unsafe { std::mem::zeroed() },
+            data_cfg: new_sdb_data_config::<T>(0),
+        }
+    }
+
+    fn db_create_or_open<P: AsRef<Path>>(
+        &mut self,
+        path: &P,
+        cfg: &DBConfig,
+        open_existing: bool,
+    ) -> Result<()> {
+        let path = path_as_cstring(path); // don't drop until init is done
+
+        // set up the splinterdb config
+        self.sdb_cfg.filename = path.as_ptr();
+        self.sdb_cfg.cache_size = cfg.cache_size_bytes as u64;
+        self.sdb_cfg.disk_size = cfg.disk_size_bytes as u64;
+        self.sdb_cfg.data_cfg = &mut self.data_cfg;
```

Comment:
This line creates a self referential struct, which is not move safe. We should either make the struct [pinned](https://doc.rust-lang.org/std/pin/struct.Pin.html), or move `self.data_cfg` to heap, e.g., Box<data_cfg>.

The bug can be reproduced by first creating a SpinterDB instance, then assign (move) it to a different variable.
The `self.sdb_cfg.data_cfg` would still point to the old address which is no longer valid.

Sorry for jumping into this pull request. I'm looking for ways to use SplinterDB in Rust and can't wait to try this wrapper :-)

Reply:
Hi Xiangpeng,

Thank you for catching this!  Would you be willing to add a commit to this PR fixing the bug?  I think for now pinning it would be most in the spirit of how this struct should be used.  But I'm open to your suggestions, since I'm not a rust expert.

Reply:
Sure I'm happy to fix the bug. It seems I don't have push permission on the pull request. Below is my patch, feel free to append it to the pull request! 

I go with the second option as pinning on the struct requires quite a bit change on how we use the wrapper, while boxing the variable is more straightforward to me (impose no runtime overhead as well). 
 

```diff
From 4857a9898cded6a8704b711e08455ea0daa292d8 Mon Sep 17 00:00:00 2001
From: Xiangpeng <me@haoxp.xyz>
Date: Mon, 15 Jan 2024 15:00:29 -0700
Subject: [PATCH] box the data_cfg to prevent self referential struct

---
 rust/splinterdb-rs/src/lib.rs | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/rust/splinterdb-rs/src/lib.rs b/rust/splinterdb-rs/src/lib.rs
index 3eab876..16fe991 100644
--- a/rust/splinterdb-rs/src/lib.rs
+++ b/rust/splinterdb-rs/src/lib.rs
@@ -18,7 +18,7 @@ pub struct DBConfig {
 pub struct SplinterDB {
     _inner: *mut splinterdb_sys::splinterdb,
     sdb_cfg: splinterdb_sys::splinterdb_config,
-    data_cfg: splinterdb_sys::data_config,
+    data_cfg: Box<splinterdb_sys::data_config>,
 }
 
 unsafe impl Sync for SplinterDB {}
@@ -192,7 +192,7 @@ impl SplinterDB
         SplinterDB {
             _inner: std::ptr::null_mut(),
             sdb_cfg: unsafe { std::mem::zeroed() },
-            data_cfg: new_sdb_data_config::<T>(0),
+            data_cfg: Box::new(new_sdb_data_config::<T>(0)),
         }
     }
 
@@ -208,7 +208,7 @@ impl SplinterDB
         self.sdb_cfg.filename = path.as_ptr();
         self.sdb_cfg.cache_size = cfg.cache_size_bytes as u64;
         self.sdb_cfg.disk_size = cfg.disk_size_bytes as u64;
-        self.sdb_cfg.data_cfg = &mut self.data_cfg;
+        self.sdb_cfg.data_cfg = self.data_cfg.as_mut();
 
         // set key bytes
         self.data_cfg.max_key_size = cfg.max_key_size as u64;
-- 
2.34.1
```